### ISSUE TITLE
Infer literal types for string concatenations and similar expressions

### ISSUE THREAD

**Issue by donaldpipowitch** [microsoft/TypeScript#13969]
**TypeScript Version:**  2.1.5

**Expected behavior:**

```ts
const foo = 'world'; // value type: 'world'
const bar = `hello ${foo}`; // value type: 'hello world'
```

**Actual behavior:**

```ts
const foo = 'world'; // value type: 'world'
const bar = `hello ${foo}`; // type: 'string'
```

**Motivation**

For constants which are created from other constants the IntelliSense is often more useful, when we see its value type instead of the "real" type.

More practical example:

```ts
const fooBreakpoint = '800px';
const fooMediaQuery = `@media(min-width: ${fooBreakpoint})`;
```

When `fooMediaQuery` is used I'd like to see `@media(min-width: 800px)` as its type instead of `string`.

**Comment by DanielRosenwasser** [comment]
> When `fooMediaQuery` is used I'd like to see `@media(min-width: 800px)` as its type instead of `string`.

This is interesting, mostly because it sounds like you are leveraging your editor's support of string literal types so that you can quickly glance at a type and get its content.

**Comment by fjmorel** [comment]
I'd like to be able to do something like this, building a key using string literals:
```typescript
type Obj = Record<"a" | "a1" |  "b" | "b1", string>;

function getKey(useA: boolean, useOne: boolean): keyof Obj {
  // [ts] Type 'string' is not assignable to type '"a" | "a1" | "b" | "b1"'.
  return (useA ? "a" : "b") + (useOne ? "1" : "");
}
```
Adding a cast after the string concatenation makes the error go away, but doesn't actually check that my key generation is still valid if I change Obj.

**Comment by ascott18** [comment]
This would be fantastic to see for indexing into objects with string literals.

``` ts
const myObj = {
 sourceName: "source",
 sourceId: 1,
 targetName: "target",
 targetId: 2,
}

function getValues(prefix: "source" | "target") {
 return { name: obj[prefix + "Name"], id: obj[prefix + "Id" }
}
```

**Comment by vdh** [comment]
It'd be nice to be able to use either string literals or unions.
```ts
const bar = "bar";
const foobar = `foo${bar}`; // "foobar"
let union: "bar" | "baz";
const foobarOrFoobaz = `foo${union}`; // "foobar" | "foobaz"
let generic: string;
const stillGeneric = `foo${generic}`; // string
```

**Comment by mortoray** [comment]
This would be helpful for string enums with a common prefix.

```
const prefix = "mymodule_"
export enum Actions {
  actionOne: prefix + "action_one"
  actionTwo: prefix + "action_two"
}
```

This is currently rejected with error `TS2553: Computed values are not permitted in an enum with string valued members.`

**Comment by vdh** [comment]
Maybe a const assertion syntax would suit this?

```ts
const bar = "bar"; // "bar"
const existingBehaviour = `foo-${bar}`; // string
const dynamic = process.env.BAR; // string | undefined
const existingBehaviourDynamic = `foo${dynamic}`; // string
const barOrBaz = Math.round(Math.random()) ? "bar" : "baz"; // "bar" | "baz"
// A const assertion syntax…?
const fooBar = `foo-${bar}` as const; // "foo-bar"
const fooBarOrBaz = `foo-${barOrBaz}` as const; // "foo-bar" | "foo-baz"
const bad = `foo-${dynamic}` as const; // type error
```

**Comment by jcalz** [comment]
this is not inferred yet, but now we have #40336 to represent string concatenation t the type level,

**Comment by bradzacher** [comment]
Worth noting that if you add `as const` this should work now with template literal types!

https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZiEMC8MDkB3EAnANgEzQG4YB6UmANwENcBXAUxigE8AHBgLnSz0IChQkWACNq2FDAAGACwa5cSACQBvBCAC+UmNQgwh0Iv3IxTpgHoB+IA

**Comment by jakebailey** [comment]
This was fixed in #53907, which I believe has enough testing for this.
